Adding a delete button to the user's created problem list

iqbalcodes6602 opened the pull request:
# Description

I have created a delet button in the loggin user's problem list so that he user can delete the problem without accessing the problem itself.

![image](https://github.com/omegaup/omegaup/assets/111554475/b96f44fc-e46a-4d66-a23d-f28e117f674d)


Fixes: #7452 

# Comments

Add any comments for the reviewers (e.g., indicating the beginning of the revision, something where the reviewer needs to pay special attention, etc.). If  there are no extra comments this section can be deleted.

# Checklist:

- [ ] The code follows the [coding guidelines](https://github.com/omegaup/omegaup/wiki/Coding-guidelines) of omegaUp.
- [ ] The tests were executed and all of them passed.
- [ ] If you are creating a feature, the new tests were added.
- [ ] If the change is large (> 200 lines), this PR was split into various Pull Requests. It's preferred to create one PR for changes in controllers + unit tests in PHPUnit,  and then another Pull Request for UI + tests in Jest, Cypress or both.

pabo99 commented on frontend/www/js/omegaup/components/problem/Mine.vue:
```diff
@@ -171,11 +172,29 @@
               </td>
+              <td class="text-center align-middle">
+                <button
+                  class="btn btn-danger"
```
Could you please normalize the style for the remove action? The other actions have a link style, so it would look better if you also display it as a link. Additionally, could you place this action in the last column?

pabo99 commented on frontend/www/js/omegaup/components/problem/Mine.vue:
```diff
@@ -171,11 +172,29 @@
+                :cancel-title="T.problemEditDeleteCancel"
+                @ok="$emit('remove', problem.alias)"
+              >
+                <p>{{ T.problemEditDeleteConfirmationMessage }} test -50</p>
```
nit:
```suggestion
                <p>{{ T.problemEditDeleteConfirmationMessage }}</p>
```

pabo99 commented on frontend/www/js/omegaup/problem/mine.ts:
```diff
@@ -67,6 +67,13 @@ OmegaUp.on('ready', () => {
+          remove: (problemAlias: string) => {
+            api.Problem.delete({ problem_alias: problemAlias })
+              .then(() => {
+                window.location.href = '/problem/mine/';
```
Here, we can display a message indicating that the problem was deleted instead of reloading the page. If the reason for reloading the page is to update the list of problems, you could call the `showProblems` function

pabo99 reviewed (commented):
I've left you some suggestions, please give a try. If one of them is not clear, please let me know.

pabo99 commented:
Extra points if you implement the feature "Remove all the problems selected", this is the place where you should add this new action:

![image](https://github.com/omegaup/omegaup/assets/3230352/1257598d-541b-4faf-8cc6-e71684f2e819)

iqbalcodes6602 commented:
hello @pabo sir,
i have change icon style to match the link and also removed page relaoding by calling the showProblems fuction and also added a succes message when the problem deleted

![image](https://github.com/omegaup/omegaup/assets/111554475/58a029f4-dbea-47e8-ae95-641721517bcd)

pabo99 commented on frontend/www/js/omegaup/components/problem/Mine.vue:
```diff
@@ -176,6 +177,26 @@
+                ok-variant="danger"
+                :cancel-title="T.problemEditDeleteCancel"
+                @ok="
+                  $emit('remove', 
+                    problem.alias,
+                    shouldShowAllProblems,
+                  )
+                "
```
We recommend pass only one argument when we emit any action to the TypeScript files. So, if we want to pass more than one calues we prefer pass an object.

iqbalcodes6602 replied:
initially i did as a object but then i saw this function in which 2 args were passed
'change-visibility': (
            selectedProblems: types.ProblemListItem[],
            visibility: number,
          ) => {
so i did the same should i change it?

pabo99 replied:
Currently, we are aiming to minimize the passing of multiple arguments. Please feel free to include the object as you originally did.

In the future, we are considering implementing a rule in the linter to disallow this practice.

iqbalcodes6602 replied:
also since i am working on this component should i change the change-visibility fucntion as well to take object instead of mupltiple args?

pabo99 replied:
Don't worry about that. It's ok if you change it in this PR or you do it in a new one.

pabo99 commented on frontend/www/js/omegaup/components/problem/Mine.vue:
```diff
@@ -16,7 +16,7 @@
     </div>
     <div class="card">
-      <h5 class="card-header">{{ T.myproblemsListMyProblems }}</h5>
+      <h5 class="card-header">{{ T.myproblemsListMyProblems }} test 100</h5>
```
Nit:
```suggestion
      <h5 class="card-header">{{ T.myproblemsListMyProblems }}</h5>
```

iqbalcodes6602 replied:
sorry i intentionally do that so that i can see the changes have been reflected or not. sorry for this, i will change it finally 
:(

iqbalcodes6602 commented:
hello @pabo99  sir,
i have done everything as you but i am not sure why the tests are failing, except that the funcationalitites are working fine.

iqbalcodes6602 commented:
Hello @pabo99 i have completed the phase 1 my username on omegaup is iqbal6602, I have also done everything in this PR as well can you assign this to me and consider this as completion of my phase 2 so that I can submit the draft proposal for for Gsoc

pabo99 commented:
Waiting for PR #7476 to be merged, in order to make more changes in this PR

pabo99 commented on frontend/www/js/omegaup/components/problem/Mine.vue:
```diff
@@ -64,6 +64,7 @@
               <option selected value="-1">{{ T.forSelectedItems }}</option>
               <option value="1">{{ T.makePublic }}</option>
               <option value="0">{{ T.makePrivate }}</option>
+              <option value="2">{{ T.wordsDelete }}</option>
```
Could you add a computed property, and then, here you can display the "Delete" option only when the selected problems can be removed?

pabo99 replied:
done!

pabo99 commented on frontend/www/js/omegaup/components/problem/Mine.vue:
```diff
@@ -176,10 +178,47 @@
                 </a>
               </td>
+              <td class="text-center align-middle">
+                <a href="#" @click.prevent="showConfirmationModal = true">
```
Could you use a button instead of an anchor?

pabo99 replied:
done!

pabo99 commented on frontend/www/js/omegaup/components/problem/Mine.vue:
```diff
@@ -176,10 +178,47 @@
+                  <font-awesome-icon :icon="['fas', 'trash']" />
+                </a>
+              </td>
+              <b-modal
```
This modal should be created only when the problem can be deleted, could you add this validation?

pabo99 replied:
done!

pabo99 commented on frontend/www/js/omegaup/components/problem/Mine.vue:
```diff
@@ -233,7 +274,9 @@ export default class ProblemMine extends Vue {
   shouldShowAllProblems = false;
   selectedProblems: types.ProblemListItem[] = [];
-  allProblemsVisibilityOption = -1;
+  allProblemsVisibilityOption = '-1';
```
Why was the data type changed in this variable? I think it's better to leave them as numbers instead of strings.

pabo99 replied:
done!

pabo99 commented on frontend/www/js/omegaup/components/problem/Mine.vue:
```diff
@@ -243,7 +286,8 @@ export default class ProblemMine extends Vue {
   onChangeVisibility(): void {
     if (
-      this.allProblemsVisibilityOption !== -1 &&
+      (this.allProblemsVisibilityOption === '1' ||
```
this function could be enhanced if you use the [`guard clause pattern`](https://www.geeksforgeeks.org/guard-clause-in-javascript/) this way:

```ts
  onChangeVisibility(): void {
    // If no problems are selected, exit the function
    if (!this.selectedProblems.length) {
      return;
    }

    // If the option is "Remove", show the confirmation modal and exit
    if (this.allProblemsVisibilityOption == 2) {
      this.showConfirmationModalDeleteAll = true;
      return;
    }
    
    // Otherwise, emit the "change-visibility" action and reset the selected problems and visibility option
    this.$emit(
      'change-visibility',
      this.selectedProblems,
      this.allProblemsVisibilityOption,
    );
    this.selectedProblems = [];
    this.allProblemsVisibilityOption = -1;
  }
```

pabo99 replied:
done!

pabo99 reviewed (commented):
Now this PR is ready for review. I left you some suggestions. 

One more thing: you need to update your branch in order to use the new paramenter in the `problems` object

pabo99 commented:
FYI: @heduenas:

This PR would be retaken

pabo99 commented on frontend/www/js/omegaup/components/problem/Mine.vue:
```diff
@@ -176,10 +178,47 @@
                 </a>
               </td>
+              <td class="text-center align-middle">
+                <a href="#" @click.prevent="showConfirmationModal = true">
```
done!

pabo99 commented on frontend/www/js/omegaup/components/problem/Mine.vue:
```diff
@@ -176,10 +178,47 @@
               </td>
+              <td class="text-center align-middle">
+                <a href="#" @click.prevent="showConfirmationModal = true">
+                  <font-awesome-icon :icon="['fas', 'trash']" />
```
done!

pabo99 commented on frontend/www/js/omegaup/components/problem/Mine.vue:
```diff
@@ -176,10 +178,47 @@
+        :cancel-title="T.problemEditDeleteCancel"
+        @ok="
+          $emit('remove-all-problems', {
+            selectedProblems: selectedProblems,
+            shouldShowAll: shouldShowAllProblems,
```
done!
